### ISSUE TITLE
Use editor backed Reflection Probes when Update Mode set to Once

### ISSUE THREAD

**Issue by jcostello** [godotengine/godot-proposals#4782]
### Describe the project you are working on

3D Environment Lighting

### Describe the problem or limitation you are having in your project

Currently, when playing a scene, the Reflection Probes bake at initialization, which is noticeable and doesn't make sense since it was previously baked in the editor.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Use the baked information from the editor when the `update_mode` is set to `once`

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

See above

### If this enhancement will not be used often, can it be worked around with a few lines of script?

You can delay the scene with a black screen but you don't know for sure how much time it will take to bake.

### Is there a reason why this should be core and not an add-on in the asset library?

Reflection probes are core

**Comment by clayjohn** [comment]
I guess we could add an option to bake ReflectionProbes in editor and then use the baked Cubemap when running. We can't just reuse the existing one in the editor as it is generated when the scene is loaded in editor and consequently won't be generated when the game is not run from the editor. 

The only major holdback I see here is that we will need to explore efficient texture compression strategies otherwise just loading the ReflectionProbe will be as slow as rendering it

**Comment by Calinou** [comment]
When using advanced 3D rendering effects, you generally want to add some kind of fade transition when loading a scene. This is also required for TAA, volumetric fog and SDFGI to initialize.

**Comment by clayjohn** [comment]
> I don't think baking reflection probes suits Godot's editing workflow. Having to bake reflection probes manually to see changes adds a lot of friction to the level design process, and requires you to ship large data along with your game (something VoxelGI and LightmapGI already require – let's not make it worse).

We could easily make it optional. ReflectionProbe could have three options (Baked, Update Once, and Update Always). Another nice thing about baking is you can resolve dependency issues in advance. Right now, ReflectionProbes draw in tree order, which means that the first reflection probe will not capture reflections in its reflection. If you bake offline, you can ensure that all objects that are in the reflection look decent when baking.

**Comment by jcostello** [comment]
@Calinou The color rect will only hide the issue but again, you dont know how much time it will take to bake.

@clayjohn I like that it solves the dependency issues. It would be a nice fix

**Comment by VantaGhost** [comment]
I think we definitely need a way to use cubemap or panorama images as reflection probe source textures, and be able to bake and save these from reflection probes. This would allow more control on a per game basis on how the game looks and performs. Some games just don't need dynamic reflection probes and would benefit greatly from the artistic control provided by the option to use a baked texture resource. Not to mention the reduced memory needed if our games could reuse cubemaps on different probes around the game world

**Comment by VantaGhost** [comment]
My personal use case would be a game that has large areas of moving level, where even "once" updated probes will be updated constantly as they move around with the level geometry, that from the perspective of the probe would not change enough to justify re-rendering. And on an artistic level, it is more important to have something reflecting that mostly resembles the game environment, than no reflections at all. But for performance reasons, the current probes are not usable in this case.

**Comment by atirut-w** [comment]
> My personal use case would be a game that has large areas of moving level, where even "once" updated probes will be updated constantly as they move around with the level geometry, that from the perspective of the probe would not change enough to justify re-rendering.

There's already an option to only update in certain cases (initial load, etc.), but yeah. Being able to bake it and edit-time will be a great addition. Most other engines already have this AFAIK.

**Comment by VantaGhost** [comment]
I thought probes only have the "once" and "always" modes. And "once" actually updates anytime the probe itself moves if I'm not mistaken. But yes, most engines support baked reflections and it's needlessly limiting for Godot not to, real-time modes will still be an option.

**Comment by bertodelrio256** [comment]
with global shader variables you can specify a cubemap and access it via your shader code. the problem is, if you want to bake a cubemap from the current scene to use, you cant do that.

**Comment by atirut-w** [comment]
That effectively does the same thing as a global reflection source, i.e. a skybox. The point of reflection probes is that you can have multiple reflection sources so you can have e.g. an indoor room that doesn't unrealistically reflect the sky for some reason.

**Comment by bluelightzero** [comment]
I wonder if it makes sense to just open up the engine to allow developers to make any type of light.

Currently, only builtin types provide light: https://docs.godotengine.org/en/stable/tutorials/3d/lights_and_shadows.html

> - From the material itself, in the form of the emission color (though it does not affect nearby objects unless baked or screen-space indirect lighting is enabled).
> - Light nodes: DirectionalLight3D, OmniLight3D and SpotLight3D.
> - Ambient light in the [Environment](https://docs.godotengine.org/en/stable/classes/class_environment.html#class-environment) or [Reflection probes](https://docs.godotengine.org/en/stable/tutorials/3d/global_illumination/reflection_probes.html#doc-reflection-probes).
> - Global illumination ([LightmapGI](https://docs.godotengine.org/en/stable/tutorials/3d/global_illumination/using_lightmap_gi.html#doc-using-lightmap-gi), [VoxelGI](https://docs.godotengine.org/en/stable/tutorials/3d/global_illumination/using_voxel_gi.html#doc-using-voxel-gi) or [SDFGI](https://docs.godotengine.org/en/stable/tutorials/3d/global_illumination/using_sdfgi.html#doc-using-sdfgi)).

A `ShadedLight3D` node could allow any arbitrary light source to be created.

The shader (I would call it `emission()`) on the `ShadedLight3D` would accept details about a fragment it is lighting (it's position, roughness, etc...), and output the needed lighting information for the light shader on the mesh to properly render it.

This could allow for all sorts of cool effects such as textured lights, unnatural or stylistic lights, custom reflection effects, custom static cubemaps. The possibilities are endless.

----

Personally I would use this to create parallax corrected cubemaps, but to do that in godot requires me to write a shader on every single object instead of just defining it once for an area I want to to apply to.

Also, it would be nice if I could capture parts of the scene's indirect lighting to a texture through script in the editor to then use for my shader. 

Reflection probes look pretty bad IMO right now. They have aliasing problems because they are rendered at runtime and I have no control over them in shaders.

**Comment by Zireael07** [comment]
@bluelightzero I think it deserves its own proposal

**Comment by Calinou** [comment]
> Personally I would use this to create parallax corrected cubemaps

This is already available with the **Box Projection** property in ReflectionProbe. It only supports box shapes, but this provides a good enough effect in most scenes (most AAA games only use box correction for their cubemaps).

A more accurate parallax correction option would require storing the depth map in a second cubemap atlas alongside the color (similar to what's been proposed in https://github.com/godotengine/godot-proposals/issues/3013). It is technically feasible (God of War Ragnarok did it on PS5 in its Fidelity mode[^1]), but expect it to demanding in terms of performance, even if the probe is baked.

[^1]: This feature isn't available in the PC port though.

> Reflection probes look pretty bad IMO right now. They have aliasing problems because they are rendered at runtime 

See https://github.com/godotengine/godot/issues/49789. You can increase the material roughness or increase reflection probe resolution to reduce aliasing.

**Comment by clayjohn** [comment]
Just to add a note on the original proposal. What was holding us back previously was the poor quality + length of time for compressing HDR textures. I.e. compressing the textures wasn't a great option, but uncompressed would take a huge amount of memory. Now that we have Betsy merged, we should be able to compress ReflectionProbes very fast. This makes it feasible to save them ahead of time and ship them with your game without blowing up the size of your game.

**Comment by jcostello** [comment]
It would be nice to know the order in which it will bake them offline

**Comment by clayjohn** [comment]
> It would be nice to know the order in which it will bake them offline

I think the simplest workflow would be to just have a button in the editor that bakes them. It would be the same as baking lightmaps, except with ReflectionProbes it would be optional.

**Comment by bluelightzero** [comment]
> I think the simplest workflow would be to just have a button in the editor that bakes them. It would be the same as baking lightmaps, except with ReflectionProbes it would be optional.

I think to get results people expect it will need to be a separate node.

It should probably have a `Reflectionmap` with `ReflectionmapProbe` just like `LightmapGI` has with `LightmapProbe`

If you had an object scene you put a refection probe on, and reused that object scene in multiple places, you might want it to be a different reflection for each instance, so putting a `Reflectionmap` in the root scene keeps the data in the right place.

The probes are just to inform the `Reflectionmap` where to sample, and which map to render.

**Comment by Jesusemora** [comment]
> I don't think baking reflection probes suits Godot's editing workflow. Having to bake reflection probes manually to see changes adds a lot of friction to the level design process, and requires you to ship large data along with your game (something VoxelGI and LightmapGI already require – let's not make it worse).

having to wait for reflection probes to bake every time a scene is opened and every time the game is run is annoying and could be worse for low end devices.

just stopping the reflections from baking when running the game I think would be enough. pressing the "bake" button should just save the generated reflections to a file so when the scene is run this file is loaded instead of having to calculate the reflections again.

as for the "large data", **how big are reflection probe textures?** all I could find was that project settings has a reflection atlas size that is set to 256, but changing it doesn't change the resolution of the reflections.
if this is 256x256 per face of a cube, 1 reflection probe would only amount to a measly 260Kb. PBR materials use 3 textures (albedo, ORM and normal) that can currently go up to 2k and 4k and more, that being 4Mb and 16Mb respectively, which times 3 is 12Mb and 48Mb, without counting any additional textures that godot supports, like emission, SSS, clearcoat, aniso, height, etc, or people who use StandardMaterial and set a different texture for roughness and metallic. textures sizes are exponential because each increase in size has to be multiplied by 4, not 2. but this also works both ways, reducing a texture size reduces its size by 3/4ths.
and reflection is just color, just one texture, It can't be **that** much.

**Comment by Calinou** [comment]
> having to wait for reflection probes to bake every time a scene is opened and every time the game is run is annoying and could be worse for low end devices.

Regarding this particular issue, a [rework of time slicing](https://github.com/godotengine/godot/pull/97773) would make it possible to greatly speed up ReflectionProbe rendering when loading a scene. By rendering all faces at once + all mipmaps in a single frame when the scene initially loads, the pop-in should be eliminated almost entirely.

**Comment by jcostello** [comment]
> We can further improve this by sorting ReflectionProbes when updating them to ensure that the probes closest to the camera are updated first

I would suggest to do it also smaller to larger, to have better artistic control and ensure the probes are baked as the artist expect.

**Comment by lander-vr** [comment]
> I would suggest to do it also smaller to larger, to have better artistic control and ensure the probes are baked as the artist expect.

I'm not sure I understand why this would give better artistic control.
As far as I know, ReflectionProbes don't capture the ambient lighting/reflections from other probes, so the baking order shouldn't have any influence on the end result. If anything, I would argue it makes more sense to update larger probes first since they are more likely to have a bigger visual impact on a scene. As an example, you'd probably want the probe covering a train station to update before all the probes in the train wagons update.

**Comment by Calinou** [comment]
> ReflectionProbes don't capture the ambient lighting/reflections from other probes, so the baking order shouldn't have any influence on the resulting bakes.

With offline baking, this could be implemented as we can resolve dependency cycles in advance. Whether this will make a big difference in most scenes is debatable though.

**Comment by lander-vr** [comment]
> Whether this will make a big visual difference in most scenes is debatable though.

It would probably provide a pretty decent increase in the quality of reflections and consistencies between reflections and the actual scene, especially when not using LightmapGI. Definitely worthwhile to explore in my opinion, it would benefit plenty of scenes.

**Comment by jcostello** [comment]
> > Whether this will make a big visual difference in most scenes is debatable though.
> 
> It would probably provide a pretty decent increase in the quality of reflections and consistencies between reflections and the actual scene, especially when not using LightmapGI. Definitely worthwhile to explore in my opinion, it would benefit plenty of scenes.

Indeed.

Also I'm working on a scene with several probes and its painful to see them baking everytime I go to the scene's tab

**Comment by Steinbeuge** [comment]
Seems there are a lot of people who want to do a lot of interesting things with reflection probes, myself included.
Maybe they just need to be exposed in a way similar to viewports?
Let users bake in depth or other data, flip faces, vertex light only and such. Let them call update when they want, for one face or all 6, and let them send the texture to one or more cubemap texture resources?
Might be something that can go hand in hand with shader templates in the future, full control over how and when cubemaps are rendered

**Comment by GTG3000** [comment]
Gonna necro this. An ability to bake a cubemap from a reflection probe would be nice. Doubly so if we could get an emission cubemap separate.

Use case I have in mind is more like the doors/windows from Atomfall - they use cubemaps of the interior/exterior to make it appear like you're looking inside/outside. For obvious reasons, those need to be baked in one scene to be used in another. 

Not quite sure how you would capture an emission cubemap right now.

**Comment by Calinou** [comment]
> Use case I have in mind is more like the doors/windows from Atomfall - they use cubemaps of the interior/exterior to make it appear like you're looking inside/outside. For obvious reasons, those need to be baked in one scene to be used in another.

This can already be achieved with a custom shader that takes a cubemap sampler as a uniform, so you can load the cubemap of your choice. This does not require the engine to be able to bake cubemaps directly, although [it's already possible to do with a script](https://github.com/Calinou/godot-cubemap-generator).

**Comment by GTG3000** [comment]
> > Use case I have in mind is more like the doors/windows from Atomfall - they use cubemaps of the interior/exterior to make it appear like you're looking inside/outside. For obvious reasons, those need to be baked in one scene to be used in another.
> 
> This can already be achieved with a custom shader that takes a cubemap sampler as a uniform, so you can load the cubemap of your choice. This does not require the engine to be able to bake cubemaps directly, although [it's already possible to do with a script](https://github.com/Calinou/godot-cubemap-generator).

Yeah, you *can* write a script (and thanks to your repo, I did make me a one-camera scene that can be dragged around the scene to take a cubemap while the game is running), but since reflection probes already capture the cubemaps for reflection purposes, it'd be nice if you could just go

```gdscript
%ReflectionProbe.update();
var newMap: CubeMap = %ReflectionProbe.get_cubemap();
```

Though on the other hand, with my very surface level understanding of C++, it looks like code to extract reflection probe data may be bigger and more complex than having a few basis constants and a sub viewport.

**Comment by unfa** [comment]
**TL;DR - I wish RenderingServer would let me at least peek at cubemap textures.**

EDIT:

As one usually does, after writing all of the following I have managed to find a possible thread to pull on to solve my problems. The Godot documentations says a `visual_scenario` contains the reflection atlas for a scene...

---

**Disclaimer: I have been trying to work this around with varying success since Godot 4 was in pre-alpha. A lot of what I write might be outdated, as it's a result of years of trying to figure something out.** 

Ideally I'd want to be able to tell when any reflection probe is still waiting to bake, and get a signal once it is ready.
Also - I would be so happy to be able to access (read/write) the cubemap texture from `RenderingServer` - with that I could implement pre-baking myself or at least poll and know when the cubemap looks good and I can show the scene to the user. 

Then I could show a progress bar on a loading screen. Right now I am doing a very silly dance and song where I render frames at low resolution and monitor for sudden changes in frame contents. I try to wait out a certain number of frames without large changes and then I consider the scene "visually stable". This is not fully reliable though, and even after I reparent the game world from the temporary viewport to the main one - the cubemaps sometimes seem to reset and need baking again... 

Reflection probes tend to be very unpredictable, so assuming a certain amount of `process_frames` per cube side is not enough to estimate how long it will take to have the scene be really ready to present to the user. It can take 3 frames per cube side, so 18 per reflection probe or more, or less - no way of knowing.

Also it seems Godot is quite clever and optimizes rendering to not update cube maps for viewports which don't end up in the final game window's output. This means that my loading screen needs to be slightly transparent for this trick to work. It's possible I am grabbing frames from the wrong place though. Viewport refresh mode `always` doesn't seem to be a workaround for this smart viewport culling.  

This problem is exasterbated when sometimes reflection probes initialize using garbage data from VRAM, and cause strong flashing, as well as broken lighting and this cascades to other reflection probes. As a result it has historically been impossible to know if any given scene will look as expected on any given machine on any given attempt.

Being able to manually review reflection probe data in editor or live in-game, save and load that data would be a huge help in assuring quality and creating custom tooling to deal with the unpredictabile nature of things.

Thanks, I hope the ref probe workflow will continue to improve, it already has done amazing progress in the past year!

**Comment by bertodelrio256** [comment]
good find with the visual scenario @unfa I too have been trying to customize the shader handling of reflection probes for years with workarounds. I currently implemented a script which captures a single cubemap from the scene and loads it into a shader global override per scene. I the use that in my character shader to do custom desaturated ambient lighting using the RADIANCE custom variable to ignore the reflection probes. By the way you actually can calculate the rendering time for probes before showing the scene: [(https://github.com/godotengine/godot-proposals/issues/11198#issuecomment-2499199784)]

Even with all this there still is a desire to be able to control reflectionprobe capture and shading. @Calinou if im not mistaken, the current plan is to create some kind of system to replace scene.glsl or make it editable in the future.

**Comment by Arazalera** [comment]
How it could work is a ReflectionProbe could a Cubemap resource (which could be redone for things like roughness layers), which if assigned, will be used, otherwise work as usual, and the cubemap can be baked to

**Comment by bertodelrio256** [comment]
> How it could work is a ReflectionProbe could a Cubemap resource (which could be redone for things like roughness layers), which if assigned, will be used, otherwise work as usual

Yeah that would be ideal. I just had a thought: what if you could also load a script or shader in each probe that applies to reflection probe cubemap right after its captured and retriggered after any recapture. Access to the stages of capture as well, like diffuse, irradiance, and specular. I dont believe specular is captured though. That way you could do processing like HSL, contrast, or even procedurally generate a texture such as a gradient.

**Comment by c64cosmin** [comment]
I am adding my perspective on this as well, I am implementing a high fidelity pipeline and while I can use the shaders to replace IRRADIANCE to sample baked lighting information from Blender3D, I am not able to do the same with reflections.
I could implement my own way of rendering the RADIANCE but I find the ReflectionProbes to be much superior than anything I could implement, the only problem I'm noticing is that I cannot push my own baked textures.

Is there any work around to this?

**Comment by bertodelrio256** [comment]
> I am adding my perspective on this as well, I am implementing a high fidelity pipeline and while I can use the shaders to replace IRRADIANCE to sample baked lighting information from Blender3D, I am not able to do the same with reflections. I could implement my own way of rendering the RADIANCE but I find the ReflectionProbes to be much superior than anything I could implement, the only problem I'm noticing is that I cannot push my own baked textures.
> 
> Is there any work around to this?

Yes, you can place a small csgsphere (*should use meshinstance sphere per hugos comment) with the normals facing in (flip faces) as a child of the probe positioned so that the reflection probe center is inside of it. Then add an unlit material to the sphere and assign a equirectangular render of your scene in blender as the albedo texture. Set the probe to “indoor” and move it slightly to update it with your image. Now the probe has both the radiance and irradiance of your custom texture. To hide the sphere at runtime, you could add it to a visuallayer that is not visible to the game camera but visible to the probe. Done!

**Comment by Calinou** [comment]
> Yes, you can place a small csgsphere with the normals facing in (flip faces) as a child of the probe positioned so that the reflection probe center is inside of it

I suggest using MeshInstance3D + SphereMesh for this instead, as it'll be faster to generate and is more detailed by default.

**Comment by Arazalera** [comment]
There could be a EnvMap resource, that (rather than a ReflectionProbe) allocates and tracks an index in the reflection atlas, a ReflectionProbe could use it, for default usage, update mode is set to Once and an inline EnvMap is set, if probes share the same EnvMap resource, they share the same map/index, for baking, update mode is set to Never, so that the EnvMap stores the baked map, when an EnvMap is loaded, it uploads it's map to the reflection atlas at an allocated index
